widget/gif.go

```diff
@@ -0,0 +1,166 @@
+package widget
+
+import (
+	"image"
+	"image/draw"
+	"image/gif"
+	"sync"
+	"time"
+
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/canvas"
+	"fyne.io/fyne/v2/storage"
+	"fyne.io/fyne/v2/widget"
+)
+
+// AnimatedGif widget shows a Gif image with many frames.
+type AnimatedGif struct {
+	widget.BaseWidget
+	min fyne.Size
+
+	src               *gif.GIF
+	dst               *canvas.Image
+	remaining         int
+	stopping, running bool
+	runLock           sync.RWMutex
+}
+
+// NewAnimatedGif creates a new widget loaded to show the specified image.
+// If there is an error loading the image it will be returned in the error value.
+func NewAnimatedGif(u fyne.URI) (*AnimatedGif, error) {
+	ret := &AnimatedGif{}
+	ret.ExtendBaseWidget(ret)
+	ret.dst = &canvas.Image{}
+	ret.dst.FillMode = canvas.ImageFillContain
+
+	if u == nil {
+		return ret, nil
+	}
+
+	return ret, ret.Load(u)
+}
+
+// CreateRenderer loads the widget renderer for this widget. This is an internal requirement for Fyne.
+func (g *AnimatedGif) CreateRenderer() fyne.WidgetRenderer {
+	return &gifRenderer{gif: g}
+}
+
+// Load is used to change the gif file shown.
+// It will change the loaded content and prepare the new frames for animation.
+func (g *AnimatedGif) Load(u fyne.URI) error {
+	g.dst.Image = nil
+	g.dst.Refresh()
+
+	read, err := storage.Reader(u)
+	if err != nil {
+		return err
+	}
+	pix, err := gif.DecodeAll(read)
+	if err != nil {
+		return err
+	}
+	g.src = pix
+	g.dst.Image = pix.Image[0]
+	g.dst.Refresh()
+
+	return nil
+}
+
+// MinSize returns the minimum size that this GIF can occupy.
+// Because gif images are measured in pixels we cannot use the dimensions, so this defaults to 0x0.
+// You can set a minimum size if required using SetMinSize.
+func (g *AnimatedGif) MinSize() fyne.Size {
+	return g.min
+}
+
+// SetMinSize sets the smallest possible size that this AnimatedGif should be drawn at.
+// Be careful not to set this based on pixel sizes as that will vary based on output device.
+func (g *AnimatedGif) SetMinSize(min fyne.Size) {
+	g.min = min
+}
+
+// Start begins the animation. The speed of the transition is controlled by the loaded gif file.
+func (g *AnimatedGif) Start() {
+	if g.isRunning() {
+		return
+	}
+	g.runLock.Lock()
+	defer g.runLock.Unlock()
+	g.running = true
```

Comment:
This doesn't need to be deferred
```suggestion
	g.running = true
	g.runLock.Unlock()
```